tasty/src/dotty/tools/tasty/TastyHeaderUnpickler.scala

```diff
@@ -118,55 +118,64 @@ class TastyHeaderUnpickler(config: UnpicklerConfig, reader: TastyReader) {
         new String(bytes, start.index, length)
       }
 
-      val validVersion = TastyFormat.isVersionCompatible(
-        fileMajor            = fileMajor,
-        fileMinor            = fileMinor,
-        fileExperimental     = fileExperimental,
-        compilerMajor        = toolMajor,
-        compilerMinor        = toolMinor,
-        compilerExperimental = toolExperimental
-      )
-
-      check(validVersion, {
-        // failure means that the TASTy file cannot be read, therefore it is either:
-        // - backwards incompatible major, in which case the library should be recompiled by the minimum stable minor
-        //   version supported by this compiler
-        // - any experimental in an older minor, in which case the library should be recompiled by the stable
-        //   compiler in the same minor.
-        // - older experimental in the same minor, in which case the compiler is also experimental, and the library
-        //   should be recompiled by the current compiler
-        // - forward incompatible, in which case the compiler must be upgraded to the same version as the file.
-        val fileVersion = TastyVersion(fileMajor, fileMinor, fileExperimental)
-        val toolVersion = TastyVersion(toolMajor, toolMinor, toolExperimental)
-
-        val compat = Compatibility.failReason(file = fileVersion, read = toolVersion)
-
-        val what = if (compat < 0) "Backward" else "Forward"
-        val signature = signatureString(fileVersion, toolVersion, what, tool = Some(toolingVersion))
-        val fix = (
-          if (compat < 0) {
-            val newCompiler =
-              if (compat == Compatibility.BackwardIncompatibleMajor) toolVersion.minStable
-              else if (compat == Compatibility.BackwardIncompatibleExperimental) fileVersion.nextStable
-              else toolVersion // recompile the experimental library with the current experimental compiler
-            recompileFix(newCompiler)
-          }
-          else upgradeFix(fileVersion)
-        )
-        signature + fix + tastyAddendum
-      })
+      checkValidVersion(fileMajor, fileMinor, fileExperimental, toolingVersion, config)
 
       val uuid = new UUID(readUncompressedLong(), readUncompressedLong())
       new TastyHeader(uuid, fileMajor, fileMinor, fileExperimental, toolingVersion) {}
     }
   }
 
   def isAtEnd: Boolean = reader.isAtEnd
+}
 
-  private def check(cond: Boolean, msg: => String): Unit = {
+object TastyHeaderUnpickler {
+
+  def check(cond: Boolean, msg: => String): Unit = {
```

Comment:
This should probably stay private:
```suggestion
  private def check(cond: Boolean, msg: => String): Unit = {
```